### ISSUE TITLE
autograd.function with `setup_context` has a number of issues with `torch.compile`

### ISSUE THREAD

**Issue by Chillee** [pytorch/pytorch#130051]
### 🐛 Describe the bug

Issue 1: If we mess up the number of arguments from `autograd.Function` we will often trace out incorrect ops into the graph. In this case, the `getitem` should be on a *tuple*, not the tensor. This can lead to very strange behavior, including CUDA IMA's.

https://pastebin.com/v8EDBsK7
<img width="892" alt="image" src="https://github.com/pytorch/pytorch/assets/6355099/0e5482c1-d0e4-40eb-8dec-f23fd0bfbaf9">

Issue 2: If I set `generate_vmap_rule` on the `autograd.Function`, it results in some kind of hang. I haven't looked more into why.

https://pastebin.com/hcJaGkNH

cc: @zou3519 @ydwu4 

### Versions

N/A

**Comment by yanboliang** [comment]
Issue 1 doesn't work at eager mode.

**Comment by Chillee** [comment]
@yanboliang Yeah it doesn't work in eager mode, but it has a much more clear error in eager mode haha.

**Comment by yanboliang** [comment]
Add assertion to prevent strange behavior at #130077

**Comment by masnesral** [comment]
> Add assertion to prevent strange behavior 

@yanboliang Should we close this one? (I'm helping with issue scrubbing)

**Comment by yanboliang** [comment]
https://github.com/pytorch/pytorch/pull/146288 would resolve this issue.

**Comment by zou3519** [comment]
still an issue
